In [2]:
import requests
import time
import math
import pandas as pd
from collections import defaultdict

# --- CONFIG ---
API_URL = "https://lalit-narayan-youtube-comment-analyzer.hf.space/analyze-batch"
HF_TOKEN = "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"

CSV_PATH = "dataset.csv"
BATCH_SIZE = 20
TOXIC_THRESHOLD = 0.5   # 🔥 you can tune this later


def load_csv_data():
    df = pd.read_csv(CSV_PATH, on_bad_lines="skip")
    df.columns = df.columns.str.lower()

    print("📊 RAW:", df.shape)

    print("Sentiment values:", df["label_sentiment"].unique()[:10])
    print("Toxicity values:", df["label_toxicity"].unique()[:10])
    print("Spam values:", df["label_spam"].unique()[:10])

    data = []

    for _, row in df.iterrows():
        try:
            text = str(row["comment_text"]).strip()

            # -------- SENTIMENT --------
            sent_raw = str(row["label_sentiment"]).strip().lower()

            if sent_raw in ["positive", "pos", "1"]:
                sentiment = "positive"
            elif sent_raw in ["negative", "neg", "-1"]:
                sentiment = "negative"
            elif sent_raw in ["neutral", "neu", "0"]:
                sentiment = "neutral"
            else:
                continue

            # -------- TOXICITY --------
            tox_raw = str(row["label_toxicity"]).strip().lower()

            if tox_raw in ["toxic", "1", "yes"]:
                toxicity = 1
            elif tox_raw in ["non-toxic", "clean", "0", "no"]:
                toxicity = 0
            else:
                continue

            # -------- SPAM --------
            spam_raw = str(row["label_spam"]).strip().lower()

            if spam_raw in ["spam", "1", "yes"]:
                spam = 1
            elif spam_raw in ["not spam", "ham", "0", "no"]:
                spam = 0
            else:
                continue

            if text == "":
                continue

            data.append({
                "text": text,
                "sentiment": sentiment,
                "toxicity": toxicity,
                "spam": spam
            })

        except:
            continue

    print(f"✅ Final usable samples: {len(data)}")
    return data


def test_all_accuracy():
    test_cases = load_csv_data()

    total = len(test_cases)

    # --- Counters ---
    correct_sentiment = 0
    correct_toxicity = 0
    correct_spam = 0

    confusion_sentiment = defaultdict(int)
    confusion_toxicity = defaultdict(int)
    confusion_spam = defaultdict(int)

    num_batches = math.ceil(total / BATCH_SIZE)

    print(f"\n🎯 Testing All Metrics ({total} samples | {num_batches} batches)")
    print("-" * 60)

    headers = {"Authorization": f"Bearer {HF_TOKEN}"}
    start_time = time.time()

    for b in range(num_batches):
        batch = test_cases[b * BATCH_SIZE:(b + 1) * BATCH_SIZE]
        payload = [{"text": case["text"]} for case in batch]

        try:
            response = requests.post(API_URL, json=payload, headers=headers, timeout=60)

            if response.status_code != 200:
                print(f"❌ Batch {b+1} Failed: {response.status_code}")
                continue

            results = response.json()

            for case, res in zip(batch, results):

                # ---------------- SENTIMENT ----------------
                pred_sent = str(res.get("sentiment", "")).lower()
                exp_sent = case["sentiment"]

                if pred_sent == exp_sent:
                    correct_sentiment += 1
                else:
                    confusion_sentiment[(exp_sent, pred_sent)] += 1

                # ---------------- TOXICITY ----------------
                pred_tox_score = res.get("toxicity", 0.0)
                pred_tox = 1 if pred_tox_score >= TOXIC_THRESHOLD else 0
                exp_tox = case["toxicity"]

                if pred_tox == exp_tox:
                    correct_toxicity += 1
                else:
                    confusion_toxicity[(exp_tox, pred_tox)] += 1

                # ---------------- SPAM ----------------
                pred_spam = 1 if res.get("is_spam", False) else 0
                exp_spam = case["spam"]

                if pred_spam == exp_spam:
                    correct_spam += 1
                else:
                    confusion_spam[(exp_spam, pred_spam)] += 1

        except Exception as e:
            print(f"🚨 Error: {e}")

    total_time = time.time() - start_time

    # --- ACCURACY ---
    acc_sent = (correct_sentiment / total) * 100 if total else 0
    acc_tox = (correct_toxicity / total) * 100 if total else 0
    acc_spam = (correct_spam / total) * 100 if total else 0

    # --- REPORT ---
    print("\n" + "="*60)
    print("📊 FINAL REPORT")

    print(f"\n🟢 Sentiment Accuracy : {acc_sent:.2f}% ({correct_sentiment}/{total})")
    print(f"🔴 Toxicity Accuracy  : {acc_tox:.2f}% ({correct_toxicity}/{total})")
    print(f"🟡 Spam Accuracy      : {acc_spam:.2f}% ({correct_spam}/{total})")

    if total > 0:
        print(f"\n⏱️ Time: {total_time:.2f}s ({total_time/total:.2f}s per comment)")
    else:
        print(f"\n⏱️ Time: {total_time:.2f}s (No valid samples)")

    print("-" * 60)

    # --- CONFUSIONS ---
    print("\n🔍 SENTIMENT CONFUSION:")
    for (exp, pred), count in sorted(confusion_sentiment.items(), key=lambda x: -x[1])[:5]:
        print(f"{exp} → {pred}: {count}")

    print("\n🔍 TOXICITY CONFUSION:")
    for (exp, pred), count in sorted(confusion_toxicity.items(), key=lambda x: -x[1])[:5]:
        print(f"{exp} → {pred}: {count}")

    print("\n🔍 SPAM CONFUSION:")
    for (exp, pred), count in sorted(confusion_spam.items(), key=lambda x: -x[1])[:5]:
        print(f"{exp} → {pred}: {count}")

    print("="*60)


if __name__ == "__main__":
    test_all_accuracy()

📊 RAW: (20071, 5)
Sentiment values: ['positive' 'neutral' 'negative']
Toxicity values: ['toxic' 'non-toxic']
Spam values: ['not spam' 'spam']
✅ Final usable samples: 20071

🎯 Testing All Metrics (20071 samples | 1004 batches)
------------------------------------------------------------
❌ Batch 5 Failed: 422
❌ Batch 29 Failed: 422
❌ Batch 54 Failed: 422
❌ Batch 58 Failed: 422
❌ Batch 72 Failed: 422
❌ Batch 90 Failed: 422
❌ Batch 103 Failed: 500
❌ Batch 104 Failed: 500
❌ Batch 127 Failed: 422
❌ Batch 128 Failed: 422
❌ Batch 134 Failed: 422
❌ Batch 147 Failed: 422
❌ Batch 153 Failed: 422
❌ Batch 156 Failed: 422
❌ Batch 158 Failed: 422
❌ Batch 160 Failed: 500
❌ Batch 162 Failed: 422
❌ Batch 168 Failed: 422
❌ Batch 171 Failed: 422
❌ Batch 172 Failed: 422
❌ Batch 178 Failed: 422
❌ Batch 180 Failed: 422
❌ Batch 181 Failed: 422
❌ Batch 182 Failed: 422
❌ Batch 184 Failed: 422
❌ Batch 189 Failed: 422
❌ Batch 208 Failed: 422
❌ Batch 212 Failed: 422
❌ Batch 217 Failed: 500
❌ Batch 218 Failed: 422


here the sentiment accuracy is weak because in our pipeline when we are absolutely sure that a particular comment is spam we output this ->  if spam_result["is_spam"]:
        return {
            "intent": "spam",
            "intent_confidence": 1.0,
            "labels": [{"label": "spam", "score": 1.0}],
            "sentiment": "neutral", "sentiment_confidence": 1.0,
            "toxicity": 0.0, "is_spam": True
        }

and we dont touch the models for any analysis for that comment , so because of this we can see that sentiment converges towards neutral in most cases and , toxicity = 0 becomes correct in most cases because most spams are non toxic in nature         

PURE SENTIMENT MODEL EVALUATION

In [6]:
import time
import math
import pandas as pd
from collections import defaultdict
from transformers import pipeline
import torch

# --- CONFIG ---
CSV_PATH = "sentiment_dataset.csv"
BATCH_SIZE = 32

device = 0 if torch.cuda.is_available() else -1

# Load model
sent_pipe = pipeline(
    "sentiment-analysis",
    model="cardiffnlp/twitter-xlm-roberta-base-sentiment", # Change this line
    device=device
)

# Adjust label mapping if necessary for the new model's output
# (e.g., bertweet might output 'POS', 'NEG', 'NEU')
label_map = {
    "LABEL_0": "negative",
    "LABEL_1": "neutral",
    "LABEL_2": "positive"
}


def load_csv_data():
    df = pd.read_csv(CSV_PATH, on_bad_lines='skip')
    df.columns = df.columns.str.lower().str.strip()

    print("📊 RAW:", df.shape)
    print("Columns:", df.columns.tolist())

    data = []

    for _, row in df.iterrows():
        try:
            text = str(row["comment"]).strip()
            sent = str(row["sentiment"]).strip().lower()

            # skip bad rows
            if text == "" or text.lower() == "nan":
                continue

            if sent not in ["positive", "negative", "neutral"]:
                continue

            data.append({
                "text": text,
                "expected": sent
            })

        except:
            continue

    print(f"✅ Final usable samples: {len(data)}")
    return data


def test_sentiment_model():
    test_cases = load_csv_data()

    total = len(test_cases)
    correct = 0

    confusion = defaultdict(int)
    low_confidence = []

    num_batches = math.ceil(total / BATCH_SIZE)

    print(f"\n🎯 Testing Raw Model ({total} samples | {num_batches} batches)")
    print("-" * 60)

    start_time = time.time()

    for b in range(num_batches):
        batch = test_cases[b * BATCH_SIZE:(b + 1) * BATCH_SIZE]
        texts = [case["text"] for case in batch]

        try:
            results = sent_pipe(texts, batch_size=BATCH_SIZE)

            for case, res in zip(batch, results):
                pred_label = label_map.get(res["label"], "neutral")
                conf = res.get("score", 0.0)

                expected = case["expected"]

                if pred_label == expected:
                    correct += 1
                else:
                    confusion[(expected, pred_label)] += 1

                if conf < 0.5:
                    low_confidence.append((case["text"], pred_label, expected, conf))

        except Exception as e:
            print(f"🚨 Error: {e}")

    total_time = time.time() - start_time
    accuracy = (correct / total) * 100 if total else 0

    # --- REPORT ---
    print("\n" + "="*60)
    print("📊 RAW MODEL REPORT")

    print(f"Accuracy: {accuracy:.2f}% ({correct}/{total})")

    if total > 0:
        print(f"Time: {total_time:.2f}s ({total_time/total:.4f}s per comment)")
    else:
        print(f"Time: {total_time:.2f}s (No valid samples)")

    print("-" * 60)

    # --- CONFUSION ---
    print("🔍 CONFUSION (Top):")
    for (exp, pred), count in sorted(confusion.items(), key=lambda x: -x[1])[:10]:
        print(f"{exp} → {pred}: {count}")

    # --- LOW CONF ---
    print("\n⚠️ LOW CONFIDENCE CASES:")
    for text, pred, exp, conf in low_confidence[:10]:
        print(f"{conf:.2f} | Pred: {pred} | Exp: {exp} | {text}")

    print("="*60)


if __name__ == "__main__":
    test_sentiment_model()

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

XLMRobertaForSequenceClassification LOAD REPORT from: cardiffnlp/twitter-xlm-roberta-base-sentiment
Key                             | Status     |  | 
--------------------------------+------------+--+-
roberta.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


📊 RAW: (6802, 2)
Columns: ['comment', 'sentiment']
✅ Final usable samples: 6797

🎯 Testing Raw Model (6797 samples | 213 batches)
------------------------------------------------------------

📊 RAW MODEL REPORT
Accuracy: 29.85% (2029/6797)
Time: 16.95s (0.0025s per comment)
------------------------------------------------------------
🔍 CONFUSION (Top):
positive → neutral: 4689
negative → neutral: 79

⚠️ LOW CONFIDENCE CASES:
0.50 | Pred: neutral | Exp: negative | Mrbeast is slowly turning into mrjigsaw
0.47 | Pred: neutral | Exp: neutral | days alone with AI
0.44 | Pred: neutral | Exp: neutral | Living underwater in a submarine for days
0.49 | Pred: neutral | Exp: positive | The last part was supposed to be the hardest and he jogged through it. Lol
0.38 | Pred: neutral | Exp: neutral | Mrbeast Aloo khaoge
0.49 | Pred: neutral | Exp: neutral | Under hour gang
0.49 | Pred: neutral | Exp: neutral | Under hour gang
0.45 | Pred: neutral | Exp: neutral | under hour gang!!
0.49 | Pred: neutra

earlier we were using ->cardiffnlp/twitter-xlm-roberta-base-sentiment

📊 RAW MODEL REPORT
Accuracy: 29.85% (2029/6797)
Time: 16.95s (0.0025s per comment)


--------------------------------------
this is for finiteautomata/bertweet-base-sentiment-analysis ,

📊 RAW MODEL REPORT
Accuracy: 76.93% (5229/6797)
Time: 16.44s (0.0024s per comment)


